### PR TITLE
[WIP] Use cupy for color conversion

### PR DESCRIPTION
**Pull request by sfriedo**
I used cupy for color conversion instead of matlab or opencv.
Seemed to work on the GPU for now, but I think this actually slows us down. (I would name some numbers, but I got an error unrelated to this due to current dataset changes it seems, so this is to be tested once #5 is finally merged)

Reason for the slowdown is most probably super frequent context switch (copy img to GPU, convert it, copy back and then the same procedure again, so 4 copy procedures for every image.)

TODO:
In case we don't want to train on the GPU (because we seem to be supporting it), we still have to use matplotlib, I didn't add a parameter for this yet.

Can we maybe do the whole augmentation on the GPU and save 2 copy procedures? Most likely, but first we need to know how slow this really is.

(I'm looking at you, #5 )

### CODE REVIEW

**Comment by fapaul** [comment]
@shorschig What is the current state of this pr? How much slower are we with these changes?

**Comment by sfriedo** [comment]
Ok, so I checked it but it doesn't work with MultiprocessIterator yet. This is a bigger problem, since the error is not on our side, it's either a cupy or a chainer problem or both, this is to be figured out. ([hint maybe](https://github.com/pytorch/pytorch/issues/2517#issuecomment-325039259), but then we have to change the Iterator itself? 🤢)

On another note, SerialIterator works, and it works faster than the current master (0.3, which is rather slow, but only due to the addition of stuff to make it multiprocessable, so I guess this is irrelevant ¯\\\_(ツ)_/¯)

**Comment by fapaul** [comment]
Can you post the error message?

**Comment by sfriedo** [comment]
If you just start the learning it errors away with something along the lines of `TypeError: Integer is expected`. But that is just a general kernel error I suppose, if you try-catch the exception inside the color conversion kernel, the printed exception is `CUDAError: Initialization failed`

**Comment by fapaul** [comment]
Ok, do you think with more work you could fix this? Otherwise, we just skip the color conversion for now and keep an eye on the test set evaluation to prevent overfitting. In the worst case, we have to work on the color conversation later.

**Comment by sfriedo** [comment]
TBH, I'm absolutely clueless at the moment. I tried to set the multiprocessing context to `spawn` like described in the issue, but I get a `RuntimeError: context has already been set`. This seems to be the issue because the context was already declared at some point before (I tried to set it inside the MultiprocessIterator to be early). I can't find the point where though (neither `set_start_method` nor `get_context` seem to appear in the chainer site-package).

TL;DR: I think it's best to defer this issue and skip it for now.
